### PR TITLE
Adds a new option to core/editor to disable code editing

### LINKED ISSUE #14647: Add Option to Disable Code Editor
**Is your feature request related to a problem? Please describe.**
We are using Gutenberg for our UF Health sites which includes over 700+ sites and many more admins who aren't technical yet want to use the platform. It would really be nice if we can disable the Code Editor for them so they can only use the blocks we've specifically built for the platform. Otherwise we run into some trouble. Right now the only option is to disable the visual editor. That made sense with the classic editor but with something like Gutenberg I think it makes more sense to also have an option to disable the code editor. It's especially helpful as we try to enforce style standards across our stack.

**Describe the solution you'd like**
Please add an option, possibly to the block_editor_settings filter, to disable the code editor.

**Describe alternatives you've considered**
Right now the only thing I'm able to do is force the visual editor on load with the 'core\edit-post' data module.

### PR DESCRIPTION
**Pull request by draganescu**
## Description
Closes #14647. This adds a new option codeEditingEnabled which the block_editor_settings filter in WP Core can set to false (default is true), case in which the post editor will have two changes reflected:

- the more tools and options menu (vertical dots menu top right) will not show the option to switch between code and visual
- the blocks themselves will not have the option for HTML editing @myleshyson I thought that'd be a good idea, as no code editing means _no code editing_ :)

## How has this been tested?

1. Set the codeEditingEnabled editor setting to false with a plugin (or hack it in edit-form-blocks.php)
2. Observe the block editor while editing a post, you cannot edit the code anymore.

## Screenshots
Screenshot 2019-04-11 at 19 10 23

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by draganescu** [comment]
I think this is the first time Travis is right.

**Comment by draganescu** [comment]
updated the failing tests, should be ok for review @gziolo :)

**Review by talldan — COMMENTED** [review]
Hey @draganescu - I found one edge case that may or may not need to be addressed. 

If I'm one of many users on a site, and I've switched my editor into `Code Editor`, I believe it's saved as a preference in local storage.

If an admin then disables the `Code Editor` using this solution as it stands, when I return my editor will still be in `Code Editor` mode. I can switch back using the 'Exit Code Editor' button, but it still gives me access that to something where the intention was to disable it.

Not sure what the right solution to solve that is. We could override the result of `getEditorMode` where it's called. That would definitely add some extra cruft to the codebase. Or we could try to dispatch an action to switch the user back.

Alternatively, it may just be something that we live with.

The only other feedback I have is that an e2e test might be nice, but I'm not sure if there's a way to toggle the editor setting before running the assertions.

File: packages/edit-post/src/components/keyboard-shortcuts/index.js [hunk 1]
```diff
@@ -56,6 +56,7 @@ class EditorModeKeyboardShortcuts extends Component {
 export default compose( [
 	withSelect( ( select ) => ( {
 		isRichEditingEnabled: select( 'core/editor' ).getEditorSettings().richEditingEnabled,
```
Comment (gziolo) [inline packages/edit-post/src/components/keyboard-shortcuts/index.js]: `isRichEditingEnabled` and `isCodeEditingEnabled` can be combined into one flag to make it all simpler:
```js
const { richEditingEnabled, codeEditingEnabled } = select( 'core/editor' ).getEditorSettings();

return {
    isEnabled: richEditingEnabled && codeEditingEnabled,
    ...
};
```

File: packages/block-editor/src/components/block-settings-menu/test/block-mode-toggle.js [hunk 2]
```diff
@@ -28,6 +28,7 @@ describe( 'BlockModeToggle', () => {
 			<BlockModeToggle
 				blockType={ { supports: { html: true } } }
 				mode="visual"
+				isCodeEditingEnabled={ true }
```
[the remark this answers is not in the document]
Comment (gziolo) [reply packages/block-editor/src/components/block-settings-menu/test/block-mode-toggle.js]: Yes, it seems like a good idea to default to `true`.

Aside:
```jsx
<BlockModeToggle isCodeEditingEnabled />
```
is equal to
```jsx
<BlockModeToggle isCodeEditingEnabled={ true } />
```

File: packages/block-editor/src/components/block-settings-menu/block-mode-toggle.js [hunk 3]
```diff
@@ -12,11 +12,15 @@ import { getBlockType, hasBlockSupport } from '@wordpress/blocks';
 import { withSelect, withDispatch } from '@wordpress/data';
 import { compose } from '@wordpress/compose';
 
-export function BlockModeToggle( { blockType, mode, onToggleMode, small = false } ) {
+export function BlockModeToggle( { blockType, mode, onToggleMode, small = false, isCodeEditingEnabled } ) {
 	if ( ! hasBlockSupport( blockType, 'html', true ) ) {
 		return null;
 	}
 
+	if ( ! isCodeEditingEnabled ) {
```
Comment (gziolo) [inline packages/block-editor/src/components/block-settings-menu/block-mode-toggle.js]: Can those two guard checks be combined into one?

**Comment by gziolo** [comment]
> If I'm one of many users on a site, and I've switched my editor into `Code Editor`, I believe it's saved as a preference in local storage.
> 
> If an admin then disables the `Code Editor` using this solution as it stands, when I return my editor will still be in `Code Editor` mode. I can switch back using the 'Exit Code Editor' button, but it still gives me access that to something where the intention was to disable it.

Does it mean that the value saves in local storage takes the precedence over what you set programmatically? The main use case is going to be that someone disables code editor with PHP code. There are two remaining tasks:
- there needs to be documentation included so people could discover this feature
- we should double check if there is any additional work necessary on WordPress core side, it's probably already handled with the following [block_editor_settings](https://github.com/WordPress/wordpress-develop/blob/17146f7192cd1c63ccde76be93d75465ccd32900/src/wp-admin/edit-form-blocks.php#L374) filter

**Comment by draganescu** [comment]
Rebased and updated with code review. Remaining:

- [  ] document the option
- [  ] resolve the loading of the setting from LS

@gziolo I only tested by using the filter `block_editor_settings` and it works simply by adding a `codeEditingEnabled` key set to `false`.

**Comment by draganescu** [comment]
Regarding documentation, @gziolo I couldn't find a list of editor settings, it appears that none of these is documented:

```
	'alignWide'              => $align_wide,
	'availableTemplates'     => $available_templates,
	'allowedBlockTypes'      => $allowed_block_types,
	'disableCustomColors'    => get_theme_support( 'disable-custom-colors' ),
	'disableCustomFontSizes' => get_theme_support( 'disable-custom-font-sizes' ),
	'disablePostFormats'     => ! current_theme_supports( 'post-formats' ),
	/** This filter is documented in wp-admin/edit-form-advanced.php */
	'titlePlaceholder'       => apply_filters( 'enter_title_here', __( 'Add title' ), $post ),
	'bodyPlaceholder'        => $body_placeholder,
	'isRTL'                  => is_rtl(),
	'autosaveInterval'       => AUTOSAVE_INTERVAL,
	'maxUploadFileSize'      => $max_upload_size,
	'allowedMimeTypes'       => get_allowed_mime_types(),
	'styles'                 => $styles,
	'imageSizes'             => $available_image_sizes,
	'richEditingEnabled'     => user_can_richedit(),
	'postLock'               => $lock_details,
	'postLockUtils'          => array(
		'nonce'       => wp_create_nonce( 'lock-post_' . $post->ID ),
		'unlockNonce' => wp_create_nonce( 'update-post_' . $post->ID ),
		'ajaxUrl'     => admin_url( 'admin-ajax.php' ),
	),
```

There is this [https://developer.wordpress.org/reference/hooks/block_editor_settings/](https://developer.wordpress.org/reference/hooks/block_editor_settings/) documentation on the fact that the filter exists :)

**Comment by draganescu** [comment]
@talldan this is my [measly solution ](https://github.com/WordPress/gutenberg/pull/14932/commits/18fe6f26520a395e62cab858a42aedfe033b198b)to the local storage pref. It does the job: if you have text in the preference but someone disables via server side filter the codeEditor then you end up in visual while your preference remains text.

**Comment by gziolo** [comment]
> Regarding documentation, @gziolo I couldn't find a list of editor settings, it appears that none of these is documented:

It doesn't sound right. We should include at least this new setting in the docs and start a section for `block_editor_settings` in Gutenberg docs as well. Other filters are documented here:
https://github.com/WordPress/gutenberg/tree/master/docs/designers-developers/developers/filters
It seems like `editor-filters.md` would be the best fit.

File: packages/edit-post/src/components/layout/index.js [hunk 4]
```diff
@@ -137,16 +137,20 @@ function Layout( {
...
+		return {
+			editorSidebarOpened: select( 'core/edit-post' ).isEditorSidebarOpened(),
+			pluginSidebarOpened: select( 'core/edit-post' ).isPluginSidebarOpened(),
+			publishSidebarOpened: select( 'core/edit-post' ).isPublishSidebarOpened(),
+			hasFixedToolbar: select( 'core/edit-post' ).isFeatureActive( 'fixedToolbar' ),
+			hasActiveMetaboxes: select( 'core/edit-post' ).hasMetaBoxes(),
+			isSaving: select( 'core/edit-post' ).isSavingMetaBoxes(),
+			isRichEditingEnabled: select( 'core/editor' ).getEditorSettings().richEditingEnabled,
+			isCodeEditingEnabled,
+			mode: isCodeEditingEnabled ? select( 'core/edit-post' ).getEditorMode() : 'visual',
```
Comment (talldan) [inline packages/edit-post/src/components/layout/index.js]: This doesn't seem too bad actually, but I wonder if it might be worth putting it on another PR. My reasoning being that I think the same kind of behaviour may also be wanted for `isRichEditingEnabled`.

Comment (gziolo) [reply packages/edit-post/src/components/layout/index.js]: Yes, both should be covered as it turns out.

Comment (draganescu) [reply packages/edit-post/src/components/layout/index.js]: so should I move this to another PR? @gziolo what do you think? I find myself not having an opinion ... :) It does introduce a new behavior and it might be good to introduce this first and then the new setting.

Comment (gziolo) [reply packages/edit-post/src/components/layout/index.js]: Let's tackle it separately for both modes. We should also do something about the case when both are disabled 💥

Comment (draganescu) [reply packages/edit-post/src/components/layout/index.js]: also about covering both `isCodeEditingEnabled` and `isRichEditingEnabled`, we should be aware of this behavior:

- rich editing defaults to whether the browser of the user supports Gutenberg, but it is also subject to the editor_settings WP filter
- currently the code editor is treated as a default last option, in case Gutenberg doesn't work. If we enable a behavior where we make both optional we could end up in a situation where the post opens in edit mode but no editor (visual or text) is available.

What should happen if neither visual nor text are enabled? We need to consider this situation and make a message sort of "Contact your Website Administrator" or whatever.

Comment (gziolo) [reply packages/edit-post/src/components/layout/index.js]: Let's discuss in a separate issue. I don't have the right answer :)

**Comment by draganescu** [comment]
@gziolo I have added a snippet of documentation about the two filterable editor settings. I am not sure about the wording ...

**Comment by draganescu** [comment]
There is aparently an option already called `isTextModeEnabled` which I accidentally found in `edit-post/src/components/header/heqder-toolbar/index.js` ... Not sure if I should rename, will investigate.

File: docs/designers-developers/developers/filters/editor-filters.md [hunk 5]
```diff
@@ -30,3 +29,24 @@ var customPreviewMessage = function() {
 wp.hooks.addFilter( 'editor.PostPreview.interstitialMarkup', 'my-plugin/custom-preview-message', customPreviewMessage );
 ```
 
+## Editor settings
+
+### `block_editor_settings`
+This is a PHP filter which is applied before sending settings to the Editor in `wp-admin/edit-form-blocks.php`.
+
+You may find details about this filter [on its WordPress Code Reference page](wp-admin/edit-form-blocks.php).
```
Comment (gziolo) [inline docs/designers-developers/developers/filters/editor-filters.md]: This link is broken. It needs to be the full URL.

File: docs/designers-developers/developers/filters/editor-filters.md [hunk 6]
```diff
@@ -30,3 +29,24 @@ var customPreviewMessage = function() {
 wp.hooks.addFilter( 'editor.PostPreview.interstitialMarkup', 'my-plugin/custom-preview-message', customPreviewMessage );
 ```
 
+## Editor settings
+
+### `block_editor_settings`
+This is a PHP filter which is applied before sending settings to the Editor in `wp-admin/edit-form-blocks.php`.
```
Comment (gziolo) [inline docs/designers-developers/developers/filters/editor-filters.md]:
```suggestion
This is a PHP filter which is applied before sending settings to the WordPress block editor.
```

File: docs/designers-developers/developers/filters/editor-filters.md [hunk 7]
```diff
@@ -30,3 +29,24 @@ var customPreviewMessage = function() {
...
+#### `richEditingEnabled`
+If it is true the user can edit the content using the visual editor.
+
+It is set by default by the WordPress function [`user_can_richedit`](https://developer.wordpress.org/reference/functions/user_can_richedit/). This function checks if the user can access the visual editor and that it’s supported by the user’s browser.
+
+
+#### `codeEditingEnabled`
+Default `true`. Sets if the user can access the code editor **in adition** to the visual editor.
+
+If set to false the user will not be able to switch between Visual and Code editor. The option in the settings menu will not be available and the keyboard shortcut for switching editor types will not fire.
```
Comment (gziolo) [inline docs/designers-developers/developers/filters/editor-filters.md]:
```suggestion
If set to `false` the user will not be able to switch to the Code Editor. The option in the settings menu will not be available and the keyboard shortcut for switching editor types will not fire.  
```

File: docs/designers-developers/developers/filters/editor-filters.md [hunk 8]
```diff
@@ -30,3 +29,24 @@ var customPreviewMessage = function() {
...
+### Available default editor settings
+
+#### `richEditingEnabled`
+If it is true the user can edit the content using the visual editor.
+
+It is set by default by the WordPress function [`user_can_richedit`](https://developer.wordpress.org/reference/functions/user_can_richedit/). This function checks if the user can access the visual editor and that it’s supported by the user’s browser.
+
+
+#### `codeEditingEnabled`
+Default `true`. Sets if the user can access the code editor **in adition** to the visual editor.
```
Comment (gziolo) [inline docs/designers-developers/developers/filters/editor-filters.md]:
```suggestion
Default `true`. Indicates whether the user can access the Code Editor **in addition** to the Visual Editor.
```

File: docs/designers-developers/developers/filters/editor-filters.md [hunk 9]
```diff
@@ -30,3 +29,24 @@ var customPreviewMessage = function() {
...
+You may find details about this filter [on its WordPress Code Reference page](wp-admin/edit-form-blocks.php).
+
+The filter will send any setting to the initialized Editor, which means any editor setting that is used to configure the editor at initialisation can be filtered by a PHP WordPress plugin before being sent. 
+
+### Available default editor settings
+
+#### `richEditingEnabled`
+If it is true the user can edit the content using the visual editor.
+
+It is set by default by the WordPress function [`user_can_richedit`](https://developer.wordpress.org/reference/functions/user_can_richedit/). This function checks if the user can access the visual editor and that it’s supported by the user’s browser.
```
Comment (gziolo) [inline docs/designers-developers/developers/filters/editor-filters.md]:
```suggestion
It is set by default to the return value of the [`user_can_richedit`](https://developer.wordpress.org/reference/functions/user_can_richedit/) function. It checks if the user can access the Visual Editor and whether it’s supported by the user’s browser.
```

File: docs/designers-developers/developers/filters/editor-filters.md [hunk 10]
```diff
@@ -30,3 +29,24 @@ var customPreviewMessage = function() {
...
+This is a PHP filter which is applied before sending settings to the Editor in `wp-admin/edit-form-blocks.php`.
+
+You may find details about this filter [on its WordPress Code Reference page](wp-admin/edit-form-blocks.php).
+
+The filter will send any setting to the initialized Editor, which means any editor setting that is used to configure the editor at initialisation can be filtered by a PHP WordPress plugin before being sent. 
+
+### Available default editor settings
+
+#### `richEditingEnabled`
+If it is true the user can edit the content using the visual editor.
```
Comment (gziolo) [inline docs/designers-developers/developers/filters/editor-filters.md]:
```suggestion
If it is `true` the user can edit the content using the Visual Editor.
```

**Review by gziolo — COMMENTED** [review]
I shared some comments on how to improve documentation. Code part looks good.

File: docs/designers-developers/developers/filters/editor-filters.md (hunk 5 above)
Comment (draganescu) [reply docs/designers-developers/developers/filters/editor-filters.md]: ugh, bad copy paste apparently

File: packages/edit-post/src/components/layout/index.js (hunk 4 above)
Comment (talldan) [inline packages/edit-post/src/components/layout/index.js]: @draganescu - just a reminder to revert this change if the plan is to tackle it separately.

**Review by gziolo — CHANGES_REQUESTED** [review]
I'm looking forward to seeing the remaining comments addressed as I'd like to ✅ this PR. It's nearly ready :)

**Comment by draganescu** [comment]
@swissspidy I am back on it starting today!

**Review by gziolo — APPROVED** [review]
Great work on this PR. Feel free to merge once Travis is green. There was some unrelated failure on one of the jobs so I restarted it.

File: packages/block-editor/src/components/block-settings-menu/block-mode-toggle.js [hunk 11]
```diff
@@ -36,10 +36,12 @@ export default compose( [
 	withSelect( ( select, { clientId } ) => {
 		const { getBlock, getBlockMode } = select( 'core/block-editor' );
 		const block = getBlock( clientId );
+		const isCodeEditingEnabled = select( 'core/editor' ).getEditorSettings().codeEditingEnabled;
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-settings-menu/block-mode-toggle.js]: We can't use `core/editor` in the block editor module. This makes this component not usable outside the post editor (widget editor, standalone editor...).

The fix is to "proxy" this config into the block editor settings (in the `EditorProvider` component) and replace this call with `select( 'core/block-editor' ).getEditorSettings()`

Anyone willing to take care of this before the next release?
